src/Middleware/RateLimiting/src/RateLimiterApplicationBuilderExtensions.cs

```diff
@@ -34,6 +38,19 @@ public static IApplicationBuilder UseRateLimiter(this IApplicationBuilder app, R
         ArgumentNullException.ThrowIfNull(app);
         ArgumentNullException.ThrowIfNull(options);
 
+        VerifyServicesAreRegistered(app);
+
         return app.UseMiddleware<RateLimitingMiddleware>(Options.Create(options));
     }
+
+    private static void VerifyServicesAreRegistered(IApplicationBuilder app)
+    {
+        var serviceProviderIsService = app.ApplicationServices.GetService<IServiceProviderIsService>();
+        if (serviceProviderIsService != null && !serviceProviderIsService.IsService(typeof(RateLimitingMetrics)))
+        {
+            throw new InvalidOperationException(Resources.FormatUnableToFindServices(
+                nameof(IServiceCollection),
+                nameof(RateLimiterServiceCollectionExtensions.AddRateLimiter)));
+        }
```

Comment:
Rate limiting middleware previously didn't have any required services. You could use `UseRateLimiter` without `AddRateLimiting`.

Now there is a require service and this validates the add services method has been called. It provides a friendly error message if not available.

In the past, we've added a marker type to DI to check this. Here I used `IServiceProviderIsService` instead to avoid adding a marker type. `IServiceProviderIsService` was added in .NET 6.

Reply:
> In the past, we've added a marker type to DI to check this. Here I used IServiceProviderIsService instead to avoid adding a marker type. IServiceProviderIsService was added in .NET 6.

I like this. Marker types have always given me a bit of an ick because they pollute the framework API and the app's DI container, IMO. This pattern is safe to use moving forward considering .NET 6+ are the actively supported versions.

Reply:
Note that it does depend on the container used to support `IServiceProviderIsService`. But looks like everyone supports it: https://github.com/dotnet/runtime/issues/53919#issuecomment-874506613

Reply:
Hmm, it's a runtime breaking change to require new services to be registered.

Reply:
Agreed. I think the pattern of skipping service verification if `IServiceProviderIsService` is missing is fine. But in this specific case, since calling `UseRateLimiter` without calling `AddRateLimiting` worked in .NET 7, it should continue to work in .NET 8.

Reply:
Rate limiting needs services now. Making it work in .NET 8 would require an `IServiceProvider` in the ctor and constructing objects manually. That wouldn't be too bad right now, but it could become more complicated in the future.

I think this is fine as a breaking change. Almost everyone will be calling `AddRateLimiter` already to configure the middleware. Some people might be using `services.Configure<RateLimiterOptions>(o => {})`, but I don't know why they'd choose to do that over the extension method.

If someone doesn't have `AddRateLimiter`, the error message makes it very clear how to fix it, which is a one-liner. And we can make a breaking change announcement about this.

Reply:
Is what I proposed above ok?